### ISSUE TITLE
Electron inserts itself into dnf.conf and breaks kernel update procedure on Fedora

### ISSUE THREAD

**Issue by lutzke** [atom/atom#12093]
### Description

DNF is the prefered package manager for the Fedora GNU/Linux distribution. Atom's dependency, Electron, inserts a line into DNF's config file, aptly named `dnf.conf`. The inserted line alters the package manager's default behaviour with regard to kernel updates, such that: if a user (such as myself) upgrades to a new kernel, and the new kernel turns out to be unstable, there is no old kernel to fall back to. 

Default behaviour for Fedora's DNF package manager is to leave no more than two old kernels around as back ups, so that if a new version of the Linux kernel has issues, a user may easily boot to the previous version, or the one before that. Atom's dependency, Electron, alters this behaviour without any notice to, or consent from, the user. After the Atom editor has been installed on Fedora, updating the kernel removes any old versions, which, again, is not supposed to happen.

This is how a **normal** `dnf.conf` looks (no package is *ever* supposed to make changes to it):
```
[main]
gpgcheck=1
installonly_limit=3
clean_requirements_on_remove=True
```
This is how my `dnf.conf` looked after Atom had mutilated it. **The last line, added by Atom, is crucial**, and affects dnf's procedure for installing kernel updates:
```
[main]
gpgcheck=1
installonly_limit=3
clean_requirements_on_remove=True
installonlypkgs=electron
```

### Steps to Reproduce

1. Install Atom on Fedora, using the official instructions provided here: https://github.com/atom/atom/

**Expected behavior:** 
I did not expect that a text editor would exhibit what I consider to be malicious behaviour: altering my package manager's configuration file without my knowledge or consent.

**Actual behavior:**
My package manager's config file **was** altered, which consequently meant that I had to spend time troubleshooting why my system wasn't behaving as expected. I also had to do this with a broken kernel, which meant doing something as simple as Alt+Tab took about ten seconds. Overall, not a pleasant experience, and not the way I would've liked to have used the last half an hour.

Thankfully, Fedora is meant for desktop use. As far as I am aware, neither CentOS nor RHEL use DNF to manage packages. Hopefully the damage this glaring oversight has caused will be limited in scope to the mere frustration of people using Fedora on their personal machines, such as myself.

### Versions
```
$ atom --version
Atom    : 1.8.0
Electron: 0.37.8
Chrome  : 49.0.2623.75
Node    : 5.10.0

$ apm --version
apm  1.10.0
npm  1.3.6
node 4.4.5
python 2.7.11
git 2.5.5
```

**Comment by ad7d** [comment]
This issue is present in Fedora 24 as well. Except, rather than leaving a single kernel during a system update, attempting to update the system with dnf resulted in: `Error: The operation would result in removing the booted kernel: kernel-core-4.5.7-300.fc24.x86_64`

```
[user@localhost dnf]$ atom --version
Atom    : 1.8.0
Electron: 0.37.8
Chrome  : 49.0.2623.75
Node    : 5.10.0
[user@localhost dnf]$ apm --version
apm  1.10.0
npm  2.15.5
node 4.4.5
python 2.7.11
git 2.7.4
```

**Comment by lutzke** [comment]
@ad7d Thank you. I have replaced "Fedora 23" with "Fedora" in my issue accordingly.

**Comment by pwnall** [comment]
It seems that `installonlypkgs` should, at the very least, include the kernel packages.

References: https://docs.fedoraproject.org/en-US/Fedora/24/html/System_Administrators_Guide/sec-Configuring_DNF_and_DNF_Repositories.html

 Also, FWIW, I'm not sure why electron needs this special behavior.

**Comment by paulcbetts** [comment]
> Atom's dependency, Electron, alters this behaviour without any notice to, or consent from, the user. 

Do you have any citations for any of this? Electron definitely doesn't touch your dnf config in any way, and afaik neither does Atom. Are you maybe using a 3rd party repo?

**Comment by lutzke** [comment]
@paulcbetts Please see my first mesage. Look under "Steps to Reproduce."

**Comment by lutzke** [comment]
@paulcbetts Please see my first mesage. Look under "Steps to Reproduce."

**Comment by lutzke** [comment]
@pwnall Thanks, I will look into this.

**Comment by paulcbetts** [comment]
@lutzke When I say "citations" I mean, "the place in the Atom or Electron source code that does this"

**Comment by lutzke** [comment]
@paulcbetts I have better things to do than comb through the source code of an editor based on a web browser, also that's a little above my skill level. I don't see why any other piece of software would add this line to my dnf.conf, do you?

**Comment by lutzke** [comment]
@pwnall What you said seems logical. However, another Fedora user has just confirmed with me that this is what the default dnf.conf looks like:

    [main]
    gpgcheck=1
    installonly_limit=3
    clean_requirements_on_remove=True

**Comment by paulcbetts** [comment]
>  I don't see why any other piece of software would add this line to my dnf.conf, do you?

I don't, but at the same time, I *have* combed through all of the source code related to Fedora packaging and Electron, and none of it edits dnf.conf

**Comment by lutzke** [comment]
@paulcbetts This is very strange. 

I was going to do a fresh install of Fedora anyway instead of the standard upgrade, to go from version 23 to 24. I'll do that right now, and then install Atom, and see if that line appears again. I'll report back with my findings.

**Comment by mehcode** [comment]
> installonlypkgs=space separated list of packages

> Here you can provide a space-separated list of packages which dnf can install, but will never update.  See the dnf.conf(5) manual page for the list of packages which are install-only by default.

> If you add the installonlypkgs directive to /etc/dnf/dnf.conf, you should ensure that you list all of the packages that should be install-only, including any of those listed under the installonlypkgs section of dnf.conf(5). In particular, kernel packages should always be listed in installonlypkgs (as they are by default), and installonly_limit should always be set to a value greater than 2 so that a backup kernel is always available in case the default one fails to boot.


I don't use Fedora.. but if I was to guess.. `dnf` itself adds atom to `installonlypkgs` when you install from an rpm file and not from a repo. From reading the documentation I would consider that correct behavior and should not result in the strange breakage you're seeing (which I can only assume that something else on your system is broken).

**Comment by thomasjo** [comment]
I just tried to reproduce this on a fresh Fedora 24 install, and was **not** able to reproduce. My `dnf.conf` remained unchanged after installing Atom via the [official RPM](https://atom.io/download/rpm).

**Comment by lee-dohm** [comment]
Marking as `more-information-needed` waiting to hear back from @lutzke.

**Comment by ad7d** [comment]
In my case, I installed Atom via unitedrpms. I don't know what the default dnf.conf looks like in 24, but I replaced that line with `installonlypkgs=kernel installonlypkg(kernel) installonlypkg(kernel-module) installonlypkg(vm)` after looking through some posts online, although that may not be what is standard.

**Comment by thomasjo** [comment]
Turns out UnitedRPMs causes this nonsense... thanks for the hint @ad7d!

Just reproduced the issue on a fresh Fedora 24 VM following the UnitedRPMs [installation instructions](https://github.com/UnitedRPMs/unitedrpms.github.io/blob/9ba834af00c3e627196d087e7f384ba32ab74464/README.md). Immediately after running `sudo dnf install atom`, I checked the contents `/etc/dnf/dnf.conf` — lo and behold, it suddenly contained the "problem entry" described above.

Since this is not caused by any wrong doing on our side, I'm closing this.
